FLT/Deformations/Algebra/InverseLimit/Basic.lean

```diff
@@ -0,0 +1,465 @@
+    CommRing (InverseLimit G f) where
+  mul_comm := mul_comm
+
+section Action
+
+instance [∀ i, Zero (G i)] [∀ i, SMulZeroClass R (G i)]
+    [∀ i j h, ZeroHomClass (T h) (G j) (G i)]
+    [∀ i j h, MulActionHomClass (T h) R (G j) (G i)] :
+    SMulZeroClass R (InverseLimit G f) where
+  smul_zero r := by
+    ext i
+    simp
+
+instance [Zero R] [∀ i, Zero (G i)] [∀ i, SMulWithZero R (G i)]
+    [∀ i j h, MulActionHomClass (T h) R (G j) (G i)]
+    [∀ i j h, ZeroHomClass (T h) (G j) (G i)] :
+    SMulWithZero R (InverseLimit G f) where
+  zero_smul x := by
+    ext i
+    simp
+
+instance [∀ i, AddZeroClass (G i)] [∀ i, DistribSMul R (G i)]
+    [∀ i j h, AddMonoidHomClass (T h) (G j) (G i)]
+    [∀ i j h, MulActionHomClass (T h) R (G j) (G i)] :
+    DistribSMul R (InverseLimit G f) where
+  smul_add r x y := by
+    ext i
+    simp
+
+instance [Monoid R] [∀ i, AddMonoid (G i)] [∀ i, DistribMulAction R (G i)]
+    [∀ i j h, DistribMulActionHomClass (T h) R (G j) (G i)] :
+    DistribMulAction R (InverseLimit G f) :=
+  have _ i j h : MulActionHomClass (T h) R (G j) (G i) := inferInstance
+  { smul_zero := smul_zero, smul_add := smul_add }
+
+instance [Monoid R] [∀ i, Monoid (G i)] [∀ i, MulDistribMulAction R (G i)]
+    [∀ i j h, MonoidHomClass (T h) (G j) (G i)]
+    [∀ i j h, MulActionHomClass (T h) R (G j) (G i)] :
+    MulDistribMulAction R (InverseLimit G f) where
+  smul_mul r x y := by
+    ext i; simp
+  smul_one r := by
+    ext i; simp
+
+instance [Semiring R] [∀ i, AddCommMonoid (G i)] [∀ i, Module R (G i)]
+    [∀ i j h, LinearMapClass (T h) R (G j) (G i)] :
+    Module R (InverseLimit G f) :=
+  have _ i j h : DistribMulActionHomClass (T h) R (G j) (G i) := inferInstance
+  {
+    add_smul r s x := by ext i; simp [add_smul],
+    zero_smul x := by ext i; simp
+  }
+
+end Action
+
+section ToComponent
+
+variable (G f) in
+def to_component (i : ι) : InverseLimit G f → G i := fun (z : InverseLimit G f) ↦ z.1 i
+
+variable (G f) in
+def to_componentₘ (i : ι) [∀ i, Group (G i)] [∀ i j h, MonoidHomClass (T h) (G j) (G i)] :
+    InverseLimit G f →* G i where
+  toFun := to_component G f i
+  map_one' := by aesop
+  map_mul' := by aesop
+
+variable (G f) in
+def to_componentᵣ (i : ι) [∀ i, Ring (G i)] [∀ i j h, RingHomClass (T h) (G j) (G i)] :
+    InverseLimit G f →+* G i where
+  toFun := to_component G f i
+  map_one' := by aesop
+  map_mul' := by aesop
+  map_zero' := by aesop
+  map_add' := by aesop
+
+variable (G f) in
+def to_componentₗ (i : ι) {R : Type*} [Ring R]
+  [∀ i, AddCommGroup (G i)] [∀ i, Module R (G i)] [∀ i j h, LinearMapClass (T h) R (G j) (G i)] :
+    InverseLimit G f →ₗ[R] G i where
+  toFun := to_component G f i
+  map_add' := by aesop
+  map_smul' := by aesop
+
+end ToComponent
+
+section Maps
+
+variable {W : Type*} {M : ι → Type*} (maps : ∀ i, M i) [∀ i, FunLike (M i) W (G i)]
+
+variable (G f) in
+class InverseSystemHom where
+  comm : ∀ i j (h : i ≤ j) w, (f i j h) (maps j w) = maps i w
```

Comment:
I'm not entirely sure we want this to be a class. Maybe this should just be a regular def.